File: product_of_array_except_self/Product_Of_Array_Except_Self.cpp
class Product_Of_Array_Except_Self {
public:
    vector<int> productExceptSelf(vector<int>& nums) {
        if (nums.empty()) {
            return vector<int>();
        }

        long left_to_right[nums.size() + 2];
        left_to_right[0] = 1;
        for (int i = 1; i < nums.size() + 1; i++) {
            left_to_right[i] = left_to_right[i - 1] * nums[i - 1];
        }
        
        long right_to_left[nums.size() + 2];
        right_to_left[nums.size() + 1] = 1;
        for (int i = nums.size(); i > 0; i--) {
            right_to_left[i] = right_to_left[i + 1] * nums[i - 1];
        }
        
        vector<int> products;
        for (int i = 0; i < nums.size(); i++) {
            products.push_back(left_to_right[i] * right_to_left[i + 2]);
        }
        return products;
    }
};

File: product_of_array_except_self/ProductOfArrayExceptSelf.java
class ProductOfArrayExceptSelf {
    public int[] productExceptSelf(int[] nums) {
        if (nums == null || nums.length < 2) {
            return new int[0];
        }
        
        long[] leftToRight = new long[nums.length];
        leftToRight[0] = nums[0];
        for (int i = 1; i < nums.length; i++) {
            leftToRight[i] = nums[i] * leftToRight[i - 1];
        }
        
        long[] rightToLeft = new long[nums.length];
        rightToLeft[nums.length - 1] = nums[nums.length - 1];
        for (int i = nums.length - 2; i >= 0; i--) {
            rightToLeft[i] = nums[i] * rightToLeft[i + 1];
        }
        
        int[] products = new int[nums.length];
        products[0] = (int)rightToLeft[1];
        products[nums.length - 1] = (int)leftToRight[nums.length - 2];
        for (int i = 1; i < nums.length - 1; i++) {
            products[i] = (int)(leftToRight[i - 1] * rightToLeft[i + 1]);
        }
        return products;
    }
}